### ISSUE TITLE
Getting a list of repos in master / template organization

### ISSUE THREAD

**Issue by htdebeer** [repobee/repobee#877]
In a plugin's command method, I get the platform API as second argument `api`. With the call `api.get_repos()` I get a list of student repositories. However, I would like to get a list of template repositories available in the master organization. I looked in the documentation of the platform API, but I seem unable to find a method to get these template repositories.

How can I get a list of template repositories in a `command` method of a plugin?

**Comment by slarse** [comment]
Hi @htdebeer,

If I understand correctly, what you need is to "discover" the available template repositories (without a user specifying e.g. `--assignments`). The way RepoBee currently works, that's not possible out-of-the-box.

First of all, the PlatformAPI does not have a notion of template/student repositories, those are higher level concepts. Also note that the PlatformAPI operates against a specific organization, which is determined when it's instantiated. In practice, when a user specifies the names of assignments, RepoBee just searches a few namespaces (template organization, target organization and current working directory) for repositories matching the names of the assignments, and uses separate API instances for the template and target orgs.

Now, discovering template repositories is inherently problematic as there's no defining characteristic of a template repository. It's _possible_ to store template repositories in the template organization, but one can also have template repositories locally, or in the target organization. While using a template organization is strongly recommended, it's not the best fit for every use case, and so it is not required.

A workaround here would be that you _require_ users to specify a template organization, and then consider all repositories in it to be the template repositories. What you'd then need is to instantiate a PlatformAPI for the template organization. That's the tricky part, as there's no easy-to-use function to instantiate the correct API implementation (different implementations require different arguments). It looks like this internally:

https://github.com/repobee/repobee/blob/de8cd774b3b6768c1ca0a6e639010f68af944cc7/src/_repobee/cli/parsing.py#L262-L284

It totally makes sense to me to make it easy to use an existing PlatformAPI to create a new one for a different organization on the same platform. Something like this (assume that `api` is the target organization API):

```python
template_org_api = api.new_api(org=template_org)
```

Thoughts on this? Perhaps you could elaborate on your use case?

**Comment by htdebeer** [comment]
Your analysis is spot on. I assumed that the "template organization" and "target organization" were of similar scope, just different uses. And thus I expected to be able to access the "template organization" and its repositories in a similar way as I can those of the "target organization".

I now understand that the "target organization" is completely controlled and managed by RepoBee, but the "template organization" is more of a data source for RepoBee, but not necessarily limited in use to RepoBee. Thus that a user could have an organization with a lot of educational repositories, some of which she wants to use with RepoBee. Then she can use it as a "template organization", but RepoBee does not require nor expect anything from this organization except having repositories in them that match the ones specified with the `--assignments` option.

In my use case, I would make a separate "template organization" for each course, and a separate "target organization" for each course instance. In my case the "template organizations" are also dedicated RepoBee organizations like the "target organizations". I would even use a single template repository for each Canvas assignment.

Anyway, what I am trying to do is the following:

- I have a `course-init` command that creates a directory and writes a RepoBee configuration file with the "target organization" and the "template organizations" set. Then, in this directory, the user runs 
- The `assignment-init` command, which starts a wizard, asks the user to:

  - select the Canvas assignment to initialize. I download all assignments from Canvas from the course.
  - select the template repository to use for the assignment. I would like to download all repositories (or their names) from the "template organization"
  
   Then the wizard creates a subdirectory for this assignment with a RepoBee config file with the Canvas assignment ID set  in the Canvas section, and the template repository set in the RepoBee section. And also a `students.lst` file matching the configuration in the Canvas assignment.
- In that assignment subdirectory, the user can then use RepoBee without having to use most command-line options anymore. Thus she can run:
 
   - `repobee repos setup`, and the `--assignments`, `--students-file` are already set in the RepoBee config file
   - `repobee repos clone`, and the `--assignments`, `--students-file` are already set  in the RepoBee config file
   - `repobee repos clone --canvas-upload-zip`, to zip the cloned repo, zip it, and upload it to the Canvas assignment. Again, the `--assignments`, `--students-file` are already set  in the RepoBee config file, as are `--canvas-course-id`, `--canvas-assignment-id`, and so on. 

  This would make using RepoBee + Canvas a lot easier because all the adminstrative plumbing is in the various RepoBee config files and data files in the course and assignment directories.

However, for this to work as I want, I would like to present the user with a list of template repositories to choose from, and for that I need access to the "template organization". I suppose it is not hard to get access to it via the git library instead of RepoBee, so I can do that as well.

**Comment by htdebeer** [comment]
So, after running `repobee canvas init-course https://url.to/course/on/canvas` and a couple of times `repobee canvas init-assignment`, you would end up with the following directory structure:

```
\JAVA101
| - repobee.ini
| - canvas-git-map.csv
| \ Assignment1-Hello_world
|  | - repobee.ini
|  | - students.lst
|  | + s32434_s35434
|  | + s45454_s33445
|  | …
|
| \ Assignment2-Fibonacci
|  | - repobee.ini
|  | - students.lst
|  | + s32434
|  | + s33445
|  | …
|
| …
```

These `repobee.ini` files inherit from their parents. The `students.lst` are fixed for each assignment, based on the assignment configuration in Canvas. Idem for the `canvas-git-map.csv`, which is based on the course configuration in Canvas.

**Comment by slarse** [comment]
> I now understand that the "target organization" is completely controlled and managed by RepoBee

Not completely, there is nothing in RepoBee stopping users (teachers or students) from adding repositories on their own. One mode of using RepoBee (the currently poorly document "discovery mode") allows students to create repositories on their own, name them whatever, and then add them to their student teams in order for RepoBee to find them.

>  "template organization" is more of a data source for RepoBee

The template organization is mostly a data source, but RepoBee can also create repositories in any organization, for example using the `repos migrate` command. There is absolutely nothing stopping a user from setting the template organization and the target organization to be the very same organization. This is in fact the default, if no template organization is specified.

> In my use case, I would make a separate "template organization" for each course, and a separate "target organization" for each course instance. In my case the "template organizations" are also dedicated RepoBee organizations like the "target organizations". I would even use a single template repository for each Canvas assignment.

This is really the recommended way to use RepoBee, and it's how we use it at KTH. One template org per course, one template repo per assignment, one target org per course round.

> Anyway, what I am trying to do is the following: [ ... ]

Right, so I think the best thing to do here is for the Canvas plugin to require the use of a template organization being present when running the `canvas assignment-init` command, and then simply fetch all repositories from the template org and present them to the user. Even if not all of them are template repos, the user should know which one is which as they set them up themselves, so that's not too much of a problem.

So, what's needed there is still just the easy way to create an API instance for the template organization.

> I suppose it is not hard to get access to it via the git library instead of RepoBee, so I can do that as well.

It will probably be tricky for you to make it work with all different platform implementations. I would defer that gnarly stuff to RepoBee, if I were you.

What I think you need to move on with this is the configuration file inheritance (working on it), and an easy way to create an API instance for the template organization. Would you agree?

**Comment by htdebeer** [comment]
Yes! As far as I now foresee, this covers my basic use case of creating and configuring a directory tree of RepoBee + Canvas managed assignments and repositories.

**Comment by slarse** [comment]
Very good! Implementing #882 is probably easier than finishing my work on the config inheritance (which turned out to be rather tricky), so I may have a go at #882 first. In any case, these are the next two issues I'm tackling, I hope to be done with both in a couple of weeks or so.

**Comment by htdebeer** [comment]
Awesome!

**Comment by slarse** [comment]
Sorry about the delay here, life's just gotten in the way the past month, it may very well have been the busiest month of my life :). This week was however the last ridiculously busy week I'll have in a while, and so I should be able to actually address these things you need within the coming week.

**Comment by htdebeer** [comment]
Take the time you need! 

We're using RepoBee and the Canvas plugin prototype now, so we encounter all kinds of issues and oddities on our end to keep us busy anyway. So this inheritance feature will be used starting next school year.

**Comment by slarse** [comment]
@htdebeer I'm slowly getting back to work on RepoBee as my free time is clearing up. A couple of FYIs for you.

* I recently extended the `PlatformAPI` with a `for_organization` method that creates a new API instance for another organization, see #904 
* The config inheritance is underway, just merged the "plumbing" (see #909), but inheritance is not yet respected by the core application. But that should also be in soon.
* The `Config` class is now actually in the docs, see https://docs.repobee.org/en/latest/repobee_plug/code.html#module-repobee_plug.config

I must admit that I've slightly lost track of what I was up to during my time away from actively working on the project, so if you feel like something is being forgotten... it very possibly is, just ping me on it :)

**Comment by slarse** [comment]
I'm going to close this issue as I'm pretty sure that it's been addressed by the addition of `PlatformAPI.for_organization`. I.e. given a PlatformAPI, you can do something like so:

```python
template_org_api = api.for_organization(template_org_name)
repos_in_template_org = template_org_api.get_repos()
```

Feel free to reopen if there's still some obstacle.
